### PR TITLE
feat: support client interceptors

### PR DESCRIPTION
**Pull request by zbigniew-mandziejewicz-3r**
Support client-side unary and streaming interceptors with similar interface to JS grpc-web interface.
Interceptors are executed by `Client.$createStreamingCall` and `Client.$createUnaryCall`.

### CODE REVIEW

**Comment by zbigniew-mandziejewicz-3r** [comment]
CLA signed.

**Comment by zbigniew-mandziejewicz-3r** [comment]
Corresponding PR for dart-lang/protobuf (https://github.com/dart-lang/protobuf/pull/389)

**Comment by zbigniew-mandziejewicz-3r** [comment]
@kevmoo could you kindly take a look?

**Comment by kevmoo** [comment]
I'll have to have @mit-mit and/or @sigurdm take a look @zbigniew-mandziejewicz-3r

**Comment by zbigniew-mandziejewicz-3r** [comment]
@kevmoo any update?

**Comment by listepo** [comment]
A very useful feature, can someone look at the PR?

**Review by mraleph — COMMENTED** [review]
Looks good to me. Thank you for your contribution and apologies for delayed response. 

I have added some feedback. Please consider addressing it.  

We are currently blocked on bringing repository in sync with our private internal fork, but once that is done we will be ready to merge (assuming that feedback is addressed).

File: test/src/client_utils.dart
Comment (mraleph) [reply test/src/client_utils.dart:96]: I see now that this functionality is used by generated protobuf code. I then suggest to rename `toResponseFuture` into override of `single`, just to make things a bit easier to understand.

```dart

  @override
  PromiseFuture<T> get single => /* ... */;
```

When I see `stream.single` it is immediately clear what semantics of that would be. However with `ResponseStream<T>.toResponseFuture` it was not clear.

File: lib/src/client/client.dart
Comment (wayne-li2) [inline lib/src/client/client.dart:37]: Question for my own understanding:

1. Why do we create a new List object instead of simply using the Iterable passed in?

2. Should we allow the user to determine what collection to pass in, instead of forcing List type?

File: lib/src/client/client.dart
Comment (wayne-li2) [inline lib/src/client/client.dart:53]: Could this for-loop be in a separate helper function, like:

```
ClientInvoker invokeInterceptors(Iterable<ClientInterceptor> interceptors)
```

you would need to make the abstract ClientInterceptor class suggested in another comment.

Comment (mraleph) [reply lib/src/client/client.dart:53]: I was thinking about suggesting to abstract this away - but decided against it because Dart does not allow easy abstraction over method invocation. `invokeInterceptors` would need to take a closure which describes which of two methods `interceptUnary` or `interceptStreaming` to invoke so invoking `invokeInterceptors` would have to look like this: 

```dart
invokeInterceptors((interceptor, method, request, options, continuation) => 
    interceptor.interceptUnary(method, request, options, continuation));
```

This is just too much boilerplate.

File: lib/src/client/client.dart
Comment (mraleph) [reply lib/src/client/client.dart:37]: 1. If you take interables and keep them for subsequent use it is a good idea to copy them into lists to prevent some strange modifications to the iterables from affecting your code. (This is also not-likely to be a performance sensitive code so copying is fine).
2. Users can pass anything right now (because we choose `Iterable` interface as a parameter). This is a bit incosistent with other cases in the gRPC code base where we force users to pass `List` (e.g. server interceptors). I don't have any strong opinion on that though.

File: lib/src/client/client.dart
Comment (wayne-li2) [reply lib/src/client/client.dart:53]: Ah dang. Thanks for the explanation -- still learning Dart nuances.

File: lib/src/client/client.dart
Comment (wayne-li2) [reply lib/src/client/client.dart:37]: Yeah I figured if we did not need to make a copy, we could just pass the Iterable as is, instead of forcing List type. But given point 1, point 2 thus makes sense.

**Comment by jairofloress** [comment]
Hi, do you know when this is going to be in master branch?
I've been doing some test based on #261 , but it throws a bizarre error of streaming with unary empty message while in master branch without the interceptor don't.
Do you know if the implementation will be the same as https://github.com/grpc/grpc-dart/pull/261#issue-355588065 ?

Thanks in advance.

**Comment by mraleph** [comment]
> Hi, do you know when this is going to be in master branch?

Hard to say, depends on many factors. We are still working on bringing our internal fork in sync with external, so we are not merging any external PRs yet. 

Also @zbigniew-mandziejewicz-3r needs to address review comments.

> Do you know if the implementation will be the same as #261 (comment) ?

The implementation is somewhat different, you can look at the code.

File: lib/src/client/interceptor.dart
Comment (zbigniew-mandziejewicz-3r) [reply lib/src/client/interceptor.dart]: Done.

File: lib/src/client/interceptor.dart
Comment (zbigniew-mandziejewicz-3r) [reply lib/src/client/interceptor.dart]: Done.

File: lib/src/client/interceptor.dart
Comment (zbigniew-mandziejewicz-3r) [reply lib/src/client/interceptor.dart]: Done.

File: test/src/client_utils.dart
Comment (zbigniew-mandziejewicz-3r) [reply test/src/client_utils.dart:99]: Renamed `.toResponseFuture()` to `.single`. Thanks for pointing it out!

File: lib/src/client/interceptor.dart
Comment (zbigniew-mandziejewicz-3r) [reply lib/src/client/interceptor.dart]: I think that invoker is shorter and easier name that still conveys information. I provided docstring in ClientInterceptor to give better context.

File: lib/src/client/client.dart
Comment (zbigniew-mandziejewicz-3r) [reply lib/src/client/client.dart]: Can this stay as it is now then?

File: lib/src/client/client.dart
Comment (zbigniew-mandziejewicz-3r) [reply lib/src/client/client.dart]: I guess this is resolved then?

**Comment by zbigniew-mandziejewicz-3r** [comment]
@mraleph @wayne-li2 Thanks for review! I've incorporated changes you requested here and updated codegen in dart-lang/protobuf.

File: lib/src/client/client.dart
Comment (zbigniew-mandziejewicz-3r) [reply lib/src/client/client.dart:37]: Done.

File: lib/src/client/client.dart
Comment (zbigniew-mandziejewicz-3r) [reply lib/src/client/client.dart:31]: Done.

File: lib/src/client/interceptor.dart
Comment (zbigniew-mandziejewicz-3r) [reply lib/src/client/interceptor.dart]: Done

File: lib/src/client/client.dart
Comment (mraleph) [reply lib/src/client/client.dart]: Yep.

**Comment by arttobe** [comment]
Any news on this pr

**Comment by mraleph** [comment]
PR is almost ready to land. However repo is not yet in a state where we are ready to land external PRs. We are still bringing it in sync with internal fork.

**Review by mraleph — APPROVED** [review]

**Comment by mraleph** [comment]
I am ready to land this. Just waiting for Travis to cycle.

**Comment by mraleph** [comment]
We are experiencing problems with Travis builds: they take forever to cycle, we have to wait until that is resolved.

**Comment by shaxbee** [comment]
Travis seems to be stuck again on scheduling jobs 😞

**Comment by mraleph** [comment]
@shaxbee We have contacted Travis support, so far did not hear much back.

**Comment by shaxbee** [comment]
@mraleph yeah I see same issues with other repositories.

**Comment by shaxbee** [comment]
@mraleph Important note - changes in dart-lang/protobuf#389 require Dart >=2.5 due to usage of `Stream.value`. I can update protobuf PR to use `Stream.fromIterable` if there is need to support earlier versions of Dart.

**Pull request merged**
